src/App.js

```diff
@@ -1,96 +1,108 @@
 /** @format */
 
 import './styles.css';
-import React, { useState, useEffect } from 'react';
+import React, { useContext } from 'react';
 import Post from './components/Post';
-import { db, auth } from './firebase';
 import { Typography } from '@material-ui/core';
 import Upload from './components/Upload';
 import Modal from './components/Modal';
+import {
+  AuthProvider,
+  AuthContext,
+} from './Firebase/context';
 
 export default function App() {
-  const [posts, setPosts] = useState([]);
-  const [username, setUsername] = useState('');
-  const [user, setUser] = useState(null);
+  const { username, posts, signup } =
+    useContext(AuthContext);
```

Comment:
We can only use `useContext` in a component that it's within a context provider. In this case, the App component isn't child of the `AuthContext` provider.

In fact, the provider is being defined in this context. So you have 3 solutions:

- Option A: Move the `useContext` to a component that is child of `App` - it cannot be used directly in `App`.
- Option B: Move the `AuthProvider` to a higher level - Add it to `src/index.js` instead of `src/App.js`
```js
<AuthProvider>
<App />
</AuthProvider>
```

Reply:
There's also an Option C: Use [Context Consumer](https://reactjs.org/docs/context.html#contextconsumer) directly in App, which gives you the context:

```js
<AuthContext.Consumer>
  {(context) => {
    return context.username ? (
      <div className="app_username">Hello 👋 {context.username}</div>
    ) : null;
  }}
</AuthContext.Consumer>
```

Reply:
Context is an advanced react feature that comes with multiple ways of using. I hope you understand the difference between each pattern, its own pros/cons. There isn't a "best" solution, it depends from case to case.